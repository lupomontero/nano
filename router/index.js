import { createElement } from '../index';


const NotFound = () => createElement('div', {
  id: 'NotFound',
  innerText: 'Not found :-(',
});


export const routerReducer = (action, state = {
  path: undefined,
  route: { path: undefined, component: NotFound },
  prev: undefined,
}) => {
  switch (action.type) {
    case 'ROUTER_PUSH':
      return {
        ...action.payload,
        prev: { path: state.path, route: state.route },
      };
  }

  return state;
};


const findRoute = (opts, path) =>
  opts.routes.find(item => item.path === path) || {
    path: undefined,
    component: NotFound,
  };


const push = (opts, props, path) => {
  window.history.pushState({}, 'Page Title', path);
  props.store.dispatch({
    type: 'ROUTER_PUSH',
    payload: {
      path,
      route: findRoute(opts, path),
    },
  });
};

export const Router = opts => component => (props) => {
  const router = {
    push: path => push(opts, props, path),
    back: () => push(
      opts,
      props,
      (props.store.getState().router.prev || {}).path || '/',
    ),
  };

  if (typeof props.store.getState().router.path === 'undefined') {
    setTimeout(() => router.push(window.location.pathname), 0);
  }

  return component({ ...props, router });
};
