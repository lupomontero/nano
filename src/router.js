import { createElement } from '.';


const NotFound = () => createElement('div', {
  className: 'not-found',
  innerText: 'Not found :-(',
});


const Loading = () => createElement('div', {
  className: 'loading',
});


const findRoute = (opts, path) =>
  opts.routes.find(item => item.path === path) || {
    path: undefined,
    component: NotFound,
  };


const push = (opts, props, path) => {
  window.history.pushState({}, 'Page Title', path);
  props.store.dispatch({
    type: 'ROUTER_PUSH',
    payload: {
      path,
      route: findRoute(opts, path),
    },
  });
};


export const Router = opts => Component => (props) => {
  const state = props.store.getState().router || {};
  const router = {
    push: path => push(opts, props, path),
    back: () => push(opts, props, (state.prev || {}).path || '/'),
  };

  if (typeof state.path === 'undefined') {
    setTimeout(() => router.push(window.location.pathname), 0);
    return Loading();
  }

  window.addEventListener('popstate', () => {
    router.push(window.location.pathname);
  });

  return Component({ ...props, router });
};

// alias
export const withRouter = Router;


export const routerReducer = (state = {
  path: undefined,
  route: {
    path: undefined,
    component: undefined,
  },
  prev: undefined,
}, action) => {
  switch (action.type) {
    case 'ROUTER_PUSH':
      return {
        ...action.payload,
        prev: { path: state.path, route: state.route },
      };
    default:
      return state;
  }
};
