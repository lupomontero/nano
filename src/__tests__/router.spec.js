import { Router, routerReducer } from '../router';

describe('Router (aka withRouter)', () => {
  it('should...', () => {
    expect(typeof Router).toBe('function');
  });
});

describe('routerReducer', () => {
  it('should return initial state en no previous state', () => {
    expect(routerReducer(undefined, {})).toMatchSnapshot();
  });

  it('should handle ROUTER_PUSH action', () => {
    const state1 = routerReducer(undefined, {
      type: 'ROUTER_PUSH',
      payload: { path: '/', route: { path: '/', component: () => {} } },
    });
    console.log(state1);
    const state2 = routerReducer(state1, {
      type: 'ROUTER_PUSH',
      payload: { path: '/some-path', route: { path: '/', component: () => {} } },
    });
    console.log(state2);
    // expect(routerReducer(undefined, {})).toMatchSnapshot();
  });
});
