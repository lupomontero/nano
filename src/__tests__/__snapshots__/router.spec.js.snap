// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routerReducer should return initial state en no previous state 1`] = `
Object {
  "path": undefined,
  "prev": undefined,
  "route": Object {
    "component": undefined,
    "path": undefined,
  },
}
`;
